zms: Initial support for 64 bit IDs

57300 opened the pull request:
Allow the ZMS API to optionally accept 64 bit IDs. A typedef `zms_id_t`
is added, so that the maximum ID width can be controlled using Kconfig.

The current ATE structure is already large enough that it is possible to
reserve 64 bits for IDs without increasing its total size (128 bits).
This makes the feature a natural, low footprint alternative to Settings,
for cases where the supported key namespace must be larger than 32 bit
but not arbitrarily large.

The ATE format does have to be altered to accommodate larger IDs, but
the default "32 bit" format is left as is. Now, the `struct zms_ate`
describes one of two supported formats, selected based on an `#ifdef`.
In the future, it may be possible to support multiple ATE formats at
runtime, in which case the structure can be turned into a union.

In the new, "64 bit" ATEs, the `offset` and `metadata` fields are moved
into a union, because they are found to be mutually exclusive. With the
old format, the same fields are in different locations, but one of them
always gets filled with a dummy value, depending on the given ATE type.
To cover both cases, use a `memset` which the compiler can optimize away
when appropriate.

The only limitation is that the new ATE format has no room for data CRC,
but an alternative integrity check can be implemented by the caller.

57300 committed:
tests: zms: Add test coverage for ZMS_ID_64BIT

* Update the "corrupt ATE" tests to work with the new ATE format.

* Add a basic test to verify support to 64 bit ZMS IDs.

* Add a `testcase.yaml` entry to cover the above points and also run
  lookup cache tests to evaluate the 64 bit hash function.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
tests: zms: Minor improvements

* Explicitly skip tests that depend on CONFIG_ZMS_LOOKUP_CACHE.

* In `testcase.yaml`, use `extra_configs` instead of `extra_args` for
  setting Kconfigs.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
zms: Initial support for 64 bit IDs

Allow the ZMS API to optionally accept 64 bit IDs. A typedef `zms_id_t`
is added, so that the maximum ID width can be controlled using Kconfig.

The current ATE structure is already large enough that it is possible to
reserve 64 bits for IDs without increasing its total size (128 bits).
This makes the feature a natural, low footprint alternative to Settings,
for cases where the supported key namespace must be larger than 32 bit
but not arbitrarily large.

The ATE format does have to be altered to accommodate larger IDs, but
the default "32 bit" format is left as is. Now, the `struct zms_ate`
describes one of two supported formats, selected by an `#if` condition.
In the future, it may be possible to support multiple ATE formats at
runtime, in which case the structure can be turned into a union.

In the new, "64 bit" ATEs, the `offset` and `metadata` fields are moved
into a union, because they are found to be mutually exclusive. With the
old format, the same fields are in different locations, but one of them
always gets filled with a dummy value, depending on the given ATE type.
To cover both cases, a `memset` is used, which should be optimized away
by the compiler when appropriate.

The only limitation is that the new ATE format has no room for data CRC,
but an alternative integrity check can be implemented by the caller.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

rghaddab commented on tests/subsys/fs/zms/testcase.yaml:
```diff
@@ -8,21 +8,29 @@ tests:
+      - CONFIG_ZMS_ID_64BIT=y
+      - CONFIG_ZMS_LOOKUP_CACHE=y
+      - CONFIG_ZMS_LOOKUP_CACHE_SIZE=64
+      - CONFIG_FLASH_SIMULATOR_DOUBLE_WRITES=y
```
Why do you need this options ?
If double writes are allowed then something is wrong.

57300 replied:
Removed.

rghaddab reviewed (commented):
Globally looks Ok, please update the documentation as well.
And the sample code

57300 commented on subsys/fs/zms/zms.c:
```diff
@@ -29,10 +29,8 @@ static int zms_ate_valid_different_sector(struct zms_fs *fs, const struct zms_at
 {
-	uint32_t hash = id;
-
 #ifdef CONFIG_ZMS_LOOKUP_CACHE_FOR_SETTINGS
```
Good point, I forgot to do that.

57300 replied:
Added.

57300 commented on subsys/fs/zms/zms.c:
```diff
@@ -1044,10 +1060,10 @@ static int zms_gc(struct zms_fs *fs)
 
 			if (gc_ate.len > ZMS_DATA_IN_ATE_SIZE) {
-				/* Copy Data only when len > 8
+				/* Only copy Data with large enough len
```
Updated.

rghaddab reviewed (dismissed):
@57300 looks good, but for sure you need to make a followup PR with documentation and sample code to show the usage of 64 bit IDs

kartben reviewed (changes requested):
ZMS documentation needs to be updated. https://docs.zephyrproject.org/latest/services/storage/zms/zms.html

Also,  the API changes should be documented in the Zephyr 4.3 migration guide

kartben commented:
Sorry, I didn't see @rghaddab comments before putting a -1. If people really feel strongly that it's ok as-is please dismiss my review, but I would really be in favour of doing everything at once

57300 commented:
No problem @kartben, thanks for the reminder. I do want to update the docs here soon, since I'm missing one approval on this PR anyway. I would appreciate another review from you then.

57300 committed:
doc: storage: zms: Update to reflect new additions

The ATE structure section is now divided into subsections describing
multiple ATE formats, including the one added for CONFIG_ZMS_ID_64BIT.
The sections about features and recommendations are updated as well.

Mentions of 32-bit key/ID are revised to account for 64-bit ID support.
Mentions of <= 8 bytes data size also had to be reworked, because this
property is tied to the chosen ATE format.

Some unrelated, minor changes are also sprinkled throughout.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 commented:
Rebased.

> ZMS documentation needs to be updated. https://docs.zephyrproject.org/latest/services/storage/zms/zms.html

Added a new commit.

> Also, the API changes should be documented in the Zephyr 4.3 migration guide

Skipped, because these changes don't require users to update their code.

kartben commented:
> > Also, the API changes should be documented in the Zephyr 4.3 migration guide
> 
> Skipped, because these changes don't require users to update their code.

Oh, right, but maybe at least the code sample should be updated as otherwise it might not necessarily be clear to people that it doesn't have to be 32-bit only anymore?

57300 commented:
> > > Also, the API changes should be documented in the Zephyr 4.3 migration guide
> > 
> > 
> > Skipped, because these changes don't require users to update their code.
> 
> Oh, right, but maybe at least the code sample should be updated as otherwise it might not necessarily be clear to people that it doesn't have to be 32-bit only anymore?

Yes, @rghaddab and I had agreed to address this in a follow-up PR.

kartben reviewed (dismissed):
please plan on adding `CONFIG_ZMS_ID_64BIT` to the release notes for Zephyr 4.3 as part of your follow-up tasks :)

57300 commented:
Rebased again. Ping @rghaddab

rghaddab commented on tests/subsys/fs/zms/src/main.c:
```diff
@@ -965,3 +967,45 @@ ZTEST_F(zms, test_zms_input_validation)
+		 * 64 bit ID to a 32 bit one somewhere.
+		 */
+		filling_id += BIT64(32);
+	}
```
Can you add as well in this test a loop where you write small data size (4 bytes in case of 64 bits) and verify that it works ?
